### PR TITLE
ADBDEV-4059-2. Support of ModifyTable node inside the correlated Subplan

### PR DESCRIPTION
**Pull request by bimboterminator1**
This PR consists of 2 commits. The first one is a revert commit, that reverts already merged changes inroduced in the patch #596. The main reason of the revert is that previously we did not consider the case when the SubPlan's subquery contained the quals, which did not have correlation parameter. In other words, all the quals we pulled up at the top of Materialize and Broadcast, what was suboptimal, because quals without correlated parameters could be applied before the Motion. Another reason is that the broadcast-materialize logic is already implemented for the *Scans code branch, and previous solution partially dublicated this code.

The second patch is gonna be a final version (don't consider Replciated distribution) of the changes. It's commit message is provided below:

When a query had a modifying command inside the correlated SubPlan, the
ModifyTable node could be rescanned for each outer tuple. That lead to
execution errors (rescan of specific nodes is not supported). This happened
because the ParallelizeCorrelatedSubplanMutator function did not expect the
ModifyTable node inside the correlated SubPlans.

This patch adds the support of the ModifyTable nodes for correlated SubPlans.
Currently, ModifyTable node can get into the SubPlan only as the part of CTE
query, therefore, it can either be wrapped in the SubqueryScan node or be
standalone, depending on the SubqueryScan being trivial or not. This
patch affects ParallelizeCorrelatedSubplanMutator function. The patch extends
the if-clause dedicated to the choice of plan nodes, that need to be
broadcasted or focused, and then materialized. The specific conditions related
to modifying operations were added. These conditions checks whether current node
is the SubqueryScan with ModifyTable just under it or current node is a
standalone ModifyTable. If condition is satisfied, node is then processed the
same way as any other Scan-type nodes. Next, the result of ModifyTable is either
broadcasted or focused depending on the target flow type. Then the result is
materialized in order to avoid rescan of underlying nodes.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/54089

File: src/backend/cdb/cdbllize.c [hunk 1]
```diff
@@ -455,9 +457,22 @@ ParallelizeCorrelatedSubPlanMutator(Node *node, ParallelizeCorrelatedPlanWalkerC
...
+	if ((IsA(node, SubqueryScan) &&
+		 IsA(((SubqueryScan *) node)->subplan, ModifyTable)) ||
+		IsA(node, ModifyTable))
+		isWritableCTE = true;
+
 	if (IsA(node, SeqScan)
 		||IsA(node, ShareInputScan)
-		||IsA(node, ExternalScan))
+		||IsA(node, ExternalScan)
+		||isWritableCTE)
```
Comment (RekGRpth) [inline src/backend/cdb/cdbllize.c]:
```suggestion
		||(IsA(node, SubqueryScan) && IsA(((SubqueryScan *) node)->subplan, ModifyTable))
		||IsA(node, ModifyTable))
```

Comment (bimboterminator1) [reply src/backend/cdb/cdbllize.c]: done

**Review by RekGRpth — DISMISSED** [review]

**Review by dnskvlnk — DISMISSED** [review]

File: src/test/regress/expected/subselect_gp.out [hunk 2]
```diff
@@ -3238,30 +3240,32 @@ with cte as
...
  Gather Motion 3:1  (slice3; segments: 3)
    ->  Seq Scan on t2
          Filter: (SubPlan 1)
          SubPlan 1  (slice3; segments: 3)
-           ->  Subquery Scan on cte
+           ->  Result
                  Filter: (t2.i = cte.i)
                  ->  Materialize
                        ->  Broadcast Motion 3:3  (slice2; segments: 3)
-                             ->  Update on t1
```
Comment (Stolb27) [inline src/test/regress/expected/subselect_gp.out:3252]: Do we really need a full set (Insert, Delete, Update) of DML operations to test? It seems, that the patch modify all of them in the same approach.

File: src/test/regress/expected/subselect_gp.out [hunk 3]
```diff
@@ -3419,27 +3429,324 @@ with cte as
...
+ count 
+-------
+     0
+(1 row)
+
+--start_ignore
+drop table if exists t3;
+NOTICE:  table "t3" does not exist, skipping
+--end_ignore
+create table t3 (i int, j int) distributed randomly;
```
Comment (Stolb27) [inline src/test/regress/expected/subselect_gp.out:3350]: Can you explain the meaning of this additional tests?

File: src/test/regress/expected/subselect_gp.out (hunk 2 above)
Comment (bimboterminator1) [reply src/test/regress/expected/subselect_gp.out:3252]: They are redundant, right.

File: src/test/regress/expected/subselect_gp.out (hunk 3 above)
Comment (bimboterminator1) [reply src/test/regress/expected/subselect_gp.out:3350]: Test that everything works for plans with SubqueryScan node, and without it.

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/54740

**Review by Stolb27 — DISMISSED** [review]

File: src/test/regress/expected/subselect_gp.out (hunk 3 above)
Comment (Stolb27) [reply src/test/regress/expected/subselect_gp.out:3350]: So it's a way to get a trivial SubqueryScan without quals that will be thrown away during setting references.

**Comment by Stolb27** [comment]
@bimboterminator1, please squash all commits besides revert commit to the only one and rebase solution to the actual `adb-6.x-dev` state.

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/54789

**Review by RekGRpth — APPROVED** [review]

**Pull request merged**